modules/dnn/src/layer_internals.hpp

```diff
@@ -157,14 +157,17 @@ struct DataLayer : public Layer
             double scale = scaleFactors[i];
             Scalar& mean = means[i];
 
-            CV_Assert(mean == Scalar() || inputsData[i].size[1] <= 4);
+            // 1d input
+            int total_c = (inputsData[i].dims >=2) ? inputsData[i].size[1]: inputsData[i].size[0];
+
+            CV_Assert(mean == Scalar() || total_c <= 4);
             if (isFP16)
                 CV_CheckTypeEQ(outputs[i].type(), CV_16SC1, "");
             else
                 CV_CheckTypeEQ(outputs[i].type(), CV_32FC1, "");
 
             bool singleMean = true;
-            for (int j = 1; j < std::min(4, inputsData[i].size[1]) && singleMean; ++j)
+            for (int j = 1; j < std::min(4, total_c) && singleMean && inputsData[i].dims >=2; ++j)
```

Comment:
> `inputsData[i].dims >=2`

Why do we need that condition?

Reply:
if I understood correctly size[1] corresponds to channels in NCHW. in case of 1d input, singleMean branch should be used (it does not iterate over the second dimension. @vpisarev @dkurt Is it right assumption?

Reply:
0.
Problem is related to changes from here: #23473
User code goes to be broken after 4.x, mostly in a silent and unexpected way.

1.
singleMean is true iff `mean[i] = X` (same value) for all `i` in [0..total_c]`.

1D doesn't guarantee that this is true (need to add separate test for that, because current tests pass just `[0,0,0,0]`).

2.
Next, there is the same problem in OpenCL code branch few lines below.

It also raises the same exception, but it is suppressed. See FIXIT comment here: https://github.com/opencv/opencv/blame/19926e2979ef049a89dd029e2231555db40c2776/modules/dnn/src/dnn.cpp#L629

3.
Einsum violates agreement of `.setInput()` API.
Test ignores the first batch number "N". Next is expected to be "C" as channels.

Accepted layouts are:
- NCHW (2d color image, dims=4)
- NCDHW (3d, e.g. "Test_TensorFlow_layers.concat_3d" test, dims=5)
- there is no 1D input expected. "NC" layout should be used instead.

There is a lack of layouts information metadata so they must be started from "NC" dimensions to keep assumption works.

BTW, there is some input with dims=3. No idea what is that. E.g. check "Test_TensorFlow_layers.l2_normalize_3d" test (`--test_debug=1` could show more).

---

So I propose:
- to revert changes here.
- to disable Einsum test in this PR
- 4.x: to add `CHECK(inputsData[i].dims, 2, "")` check (for both CPU and OpenCL code)
- 4.x: to add check in `.setInput()` to "fail fast"
- 4.x: to fix Einsum tests and use "NC" layout to avoid violation of `.setInput()` assumptions.

Reply:
This feature if not very flexible, I agree. I think that adding a check that input has at least 2 dimensions is a good option because this initially designed for image processing networks that usually apply mean subtraction as a preprocessing step. I have not seen much models with 1D input that do the same. Usually, their input data is much more complicated (text tokens, audio sampling) and mean processing is not used.

Reply:
Can we add also a flag such as `bool preprocessing` which indicates if mean/scale were actually specified but not initialized by default with zero/ones?

Reply:
I disabled cumsum tests and created an issue for it. Let's "merge" merge PR and then try to handle the issue separately.  I prefer to have dedicated PRs for significant changes to not hide it in merge commit with conflicts.